### ISSUE TITLE
mysql cluster import_error waits for a lock

### ISSUE THREAD

**Issue by Smithx10** [apache/airflow#15724]
A user of ours has a DAG named gc_test.py that doesn't work and resulted in an "import_error".  The Percona cluster 8 peers in the cluster all had many queries stuck on  "wsrep: replicating and certifying write set(-1)".    151 filling up the available connections.  Airflow is version 1.10

```
mysql> show processlist;
+--------+-----------------+---------------------+------------+---------+--------+-------------------------------------------------+------------------+-----------+---------------+
| Id     | User            | Host                | db         | Command | Time   | State                                           | Info             | Rows_sent | Rows_examined |
+--------+-----------------+---------------------+------------+---------+--------+-------------------------------------------------+------------------+-----------+---------------+
|      1 | system user     |                     | NULL       | Sleep   | 348636 | wsrep aborter idle                              | NULL             |         0 |             0 |
|      2 | system user     |                     | NULL       | Query   | 112590 | wsrep: committed write set (12836130)           | NULL             |         0 |             0 |
|      8 | event_scheduler | localhost           | NULL       | Daemon  | 348593 | Waiting on empty queue                          | NULL             |         0 |             0 |
| 235621 | airflow_user    | 10.91.194.171:38833 | airflow_db | Query   | 112589 | wsrep: replicating and certifying write set(-1) | COMMIT           |         0 |             0 |
| 235686 | airflow_user    | 10.91.194.171:35165 | airflow_db | Query   | 112307 | wsrep: replicating and certifying write set(-1) | COMMIT           |         0 |             0 
```

I logged into all the nodes and ran: "SHOW ENGINE INNODB STATUS;" and found the following useful information.

```
---TRANSACTION 30007580, ACTIVE 113802 sec starting index read
mysql tables in use 1, locked 1
LOCK WAIT 2 lock struct(s), heap size 1136, 1 row lock(s)
MySQL thread id 1, OS thread handle 6, query id 212741036 Applying batch of row changes (delete)
------- TRX HAS BEEN WAITING 113802 SEC FOR THIS LOCK TO BE GRANTED:
RECORD LOCKS space id 36 page no 4 n bits 80 index PRIMARY of table `airflow_db`.`*import_error*` trx id 30007580 lock_mode X locks rec but not gap waiting
Record lock, heap no 8 PHYSICAL RECORD: n_fields 6; compact format; info bits 0
 0: len 4; hex 8045aba9; asc  E  ;;
 1: len 6; hex 000001c9dd12; asc       ;;
 2: len 7; hex 81000000860110; asc        ;;
 3: SQL NULL;
 4: len 30; hex 2f7573722f6c6f63616c2f616972666c6f772f646167732f67635f746573; asc /usr/local/airflow/dags/gc_tes; (total 34 bytes);
 5: len 30; hex 6e616d6520276372656174655f6f6c705f636b65636b5f64616727206973; asc name 'create_olp_ckeck_dag' is; (total 42 bytes);
```

**Comment by potiuk** [comment]
I am afraid Airflow does not work with multi-master replication. You can have at most one master where everything connect to. Airflow's DB is usually not very busy, it's there to just store meta-data of the tasks, so it does not need replicated DB to work usually.

**Comment by potiuk** [comment]
Airflow also makes heavy use of locks, so it's lilely in this case the multi-master replication caused some replication dead-locks.

**Comment by Smithx10** [comment]
@potiuk ,

Thank you for this information, I'm not sure this is documented on the Airflow Website might worth adding it... I wouldn't mind adding the PR.

I currently use HAProxy in front of the Cluster,  would it be ok to use wsrep still,  but maybe change the balance algorithm to "first" ?

**Comment by potiuk** [comment]
Difficult to say. I do not know too much about your mylti-master solution . The scheduler HA design specifically excluded the multii-master replication  https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=103092651 and only supports 'regular' db  https://airflow.apache.org/docs/apache-airflow/stable/scheduler.html - maybe that could be a great first contibution to make it clearer ?

**Comment by potiuk** [comment]
Fantastic! Indeed it looks like this might be the very reason !

**Comment by Smithx10** [comment]
@potiuk,

I made a bit of progress today.  I forked the percona-healthcheck script, ported it to go, and added raft.  It still needs to be cleaned up,  Logging, Config, Documentation,  but we will be testing this week to see how it works.

In a quick summary,  this healthchecker runs on 9200 and only informs HAProxy to route to a Synced MySQL node, and he Raft Elected leader.  If a Mysql node is unhealthy the raft cluster leader with initiate a transfer.   This should make it so we don't write to Non Sync'd nodes and more than 1 at a time.

I attempted using stick-tables, backup, and a few other haproxy settings, but ran into an issue where I would need to use a VIP so that the routing table would be the same cross nodes and we would always route the same. 

https://github.com/olafz/percona-clustercheck/compare/master...Smithx10:golang-raft?expand=1
